File: raspberry-pi-ios-client/RaspberryPiClient/Data/Infrastructure/Foxglove/Topic/FoxgloveClientAdvertise.swift
//
//  FoxgloveClientAdvertise.swift
//  RaspberryPiClient
//
//  Created by Shusuke Ota on 2025/03/20.
//

import Foundation

/// Client Advertise メッセージ（クライアント → サーバー）
/// クライアントが publish するチャンネルをサーバーに通知する
/// Ref: https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#client-advertise
struct FoxgloveClientAdvertise: Codable {
    let op: FoxgloveMessageOperation
    let channels: [FoxgloveClientAdvertiseChannel]

    init(op: FoxgloveMessageOperation = .advertise, channels: [FoxgloveClientAdvertiseChannel]) {
        self.op = op
        self.channels = channels
    }

    /// JSON 文字列にエンコードする
    func toJsonString() -> String? {
        guard let jsonData = try? JSONEncoder().encode(self),
              let jsonString = String(data: jsonData, encoding: .utf8) else {
            return nil
        }
        return jsonString
    }
}

/// Client Advertise メッセージ内のチャンネル情報（クライアント → サーバー）
/// Ref: https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#client-advertise
struct FoxgloveClientAdvertiseChannel: Codable {
    let id: UInt32
    let topic: String
    let encoding: String
    let schemaName: String
    let schema: String?
    let schemaEncoding: String?
}

File: raspberry-pi-android-client/app/src/main/java/com/shusuke/raspberry_pi_client/data/infrastructure/foxglove/topic/FoxgloveClientAdvertise.kt
package com.shusuke.raspberry_pi_client.data.infrastructure.foxglove.topic

import com.shusuke.raspberry_pi_client.data.infrastructure.foxglove.common.FoxgloveMessageOperation
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Client Advertise メッセージ（クライアント → サーバー）
 *
 * クライアントが publish するチャンネルをサーバーに通知する。
 * Ref: https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#client-advertise
 */
@Serializable
data class FoxgloveClientAdvertise(
    val op: FoxgloveMessageOperation = FoxgloveMessageOperation.Advertise,
    val channels: List<FoxgloveClientAdvertiseChannel> = emptyList(),
)

/**
 * Client Advertise メッセージ内のチャンネル情報（クライアント → サーバー）
 *
 * Ref: https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#client-advertise
 */
@Serializable
data class FoxgloveClientAdvertiseChannel(
    val id: UInt,
    val topic: String,
    val encoding: String = "json",
    @SerialName("schemaName")
    val schemaName: String,
    val schema: String? = null,
    @SerialName("schemaEncoding")
    val schemaEncoding: String? = null,
)
